ARCHITECTURE.md

```diff
@@ -0,0 +1,95 @@
+# ink!ternals
+
+
+This document describes the architecture of ink!. The information
+here targets those who want to understand or modify the inner
+workings of this project.
+
+In general we treat documentation as a first-class citizen.
+All crates mentioned below should be documented really well.
+You can find the crate documentation on docs.rs or for our
+`master` branch under GitHub pages. So for `ink_lang` e.g.:
+
+* [https://docs.rs/ink_lang/latest/ink_lang](https://docs.rs/ink_lang/latest/ink_lang) (latest)
+* [https://paritytech.github.io/ink/ink_lang](https://paritytech.github.io/ink/ink_lang) (`master`)
+
+ink! is composed of a number of crates that are all found in the
+`crates/` folder. On a high-level those can be grouped as:
+
+* `lang`: The ink! language itself.
+* `allocator`: The allocator for dynamic heap space allocation in a contract.
+* `engine`: An off-chain testing engine, it simulates a blockchain
+   environment and allows mocking specified conditions.
+* `env`: Exposes environmental functions, like information about the caller
```

Comment:
Probably worth mentioning that this is the interface to the Contracts pallet as well. Could be done later in that crate diagram I proposed